### ISSUE TITLE
Display raw password and QR code at same time.

### ISSUE THREAD

**Issue by frs2022** [danmarsden/moodle-mod_attendance#713]
Dear All, 

We face issue, we have upgrade Moodle to 4.1 version, we notes the password in attendance activity dose not  shown and not rotate as OR code as shown in attached file:
![Passwrod not shown or rotate ](https://github.com/danmarsden/moodle-mod_attendance/assets/117255482/106f3ee5-b667-4405-9965-ccdcb14bf62c)

Before we upgrade the Moodle it was rotate and shown:
![Password rotate ](https://github.com/danmarsden/moodle-mod_attendance/assets/117255482/7a7b23d9-6c1a-467a-83b7-4daceff12476)

Can you please advise us 

Thank you for you support

**Comment by danmarsden** [comment]
this was an intentional change - displaying the password on the same page is ugly and it's a bad user experience asking students to enter a long password that is hard to read within a short time frame - they should be using the QR code scanning process.

Ideally it would be good to understand "Why" you want the password to be displayed? - it's also been discussed here:
https://moodle.org/mod/forum/discuss.php?d=443512

**Comment by frs2022** [comment]
Dear Danmarsden, 

Thank you for your reply and support. 

Some classes and lecture halls are too big and wide for some students to scan the code (students' device compatibility). Hence, the password effectively substitutes code scanning when their cameras fail to capture the code. In our context, the rotated passwords and QR codes are needed.

**Comment by danmarsden** [comment]
thanks - happy for someone to add this back in a Pull request, but it needs to display nicely - not just tacked onto the bottom of the page.

**Comment by zmartinovic** [comment]
I agree it would be nice for teachers to have an option to display password even when rotating QR code. It can be one more checkbox in add session options. This would also allow students to mark attendance using mobile app.

**Comment by mdorferf** [comment]
Dear Danmarsden,
We would also be delighted if this function were to be implemented again. We suspect that some student devices cannot read the QR code correctly. Perhaps there is a connection with this issue: https://github.com/danmarsden/moodle-mod_attendance/issues/746
Thank you in advance for your support!

**Comment by danmarsden** [comment]
@mdorferf - personally I think it would be better to analyse what's going on with the student devices you mention - identifying and fixing that would be more useful to the students than giving them a long jumbled password they have to enter on a mobile device keyboard.

If you have internal development capabilities, pull requests are always welcome. If you would like to fund my time to investigate/fix/improve the plugin feel free to reach out privately.

**Comment by mdorferf** [comment]
@danmarsden Thank you for your reply! We have now been able to reproduce it: If students scan the first generated QR code but then take too long to log in (SSO login), then the self recording no longer works. When scanning the second generated QR code, self-recording also does not work.

**Comment by danmarsden** [comment]
Thanks @mdorferf - sounds like the process may need improving so that the user is prompted to scan again - once they have logged in. IMO it would be better to improve that than force them to enter a long password.

**Comment by mdorferf** [comment]
Thanks @danmarsden 
In the meantime, we have carried out several tests on different mobile devices, accounts (SSO login & manual account) and browsers (Firefox, Google Chrome, Edge, Safari). We were able to minimize the errors with the following workarounds: 

- Log in to Moodle on the device before scanning the QR-Code
If the error occurs again: 
- Change browser
- Delete the browser cache

Based on the tests, we suspect that the error occurs because of the browser cache. After internal discussions, we also agree that a long password together with the QR code is bad for usability reasons. 
From our point of view, it would be great to fix the browser cache issue and optionally display a shortened rotating password for people without a QR code scanner (if possible).
Thank you for your support!

**Comment by danmarsden** [comment]
Thanks @mdorferf - I'm not sure about the display of a shortened rotating password - sounds like a lot of work to implement a secondary rotating password mechanism - better to use the same password but have some form of control to make it genrate simpler passwords - there's no benefit in having a complex password embedded in the QR code and then a simple password being displayed to the end user.

But - the browser cache problem sounds interesting - surely there's a redirect somewhere, or some better error messaging we can display when that occurs and just ask the user to "scan again"

**Comment by mdorferf** [comment]
Thanks for your quick reply @danmarsden and your opinion on implementing a short password!
I think it would be very useful if users were shown a meaningful error message. Simply rescan won't help our users if there are browser cache issues because they'll be stuck in a kind of loop.
Perhaps a solution would be to a) fix the browser cache issue and/ or b) show users a better error message, e.g. 
Before scanning the QR code again, please use a different browser and log in to your Moodle page or clear the browser cache.

**Comment by danmarsden** [comment]
yeah - any chance you could get some screenshots or screen recording that show what the end-user gets when that browser caching issue occurs?

**Comment by mdorferf** [comment]
@danmarsden For sure, here are the screenshots from my tests:
Screenshot 1: 
This is what the user gets when he/she scans the QR code but the login was not fast enough. 
The same error message is displayed when the QR code is scanned again.
Screenshot 2: 
This is the description of the error message.
Screenshot 3: 
This is what the user gets when he/she clicks on the “Next” button.

![Image](https://github.com/user-attachments/assets/27e3f558-6977-48d1-92ae-a8e0d4b4fe99)
![Image](https://github.com/user-attachments/assets/4380b7a1-f18a-4643-bfb2-c1da89559d9d)
![Image](https://github.com/user-attachments/assets/f2c2f007-9695-4c67-a615-e8ec2cb5f3e5)

**Comment by danmarsden** [comment]
I've just improved the lang string a bit to make it clear the user should scan the QR Code again. I've also created a very basic help doc in the 405 moodledocs wiki for that error that explains it can happen when the user is not already logged into the site when they scan the QR code.

I think we should hide the "enter password" field on the main attendance tracking page when the QR code is in use and put some text there to say "please scan the QR code provided by your teacher" which might help a bit more.

But - that weird issue you're seeing where the users scan twice and it still doesn't let them through sounds interesting to me - do you know if the machine the teacher is using to display the QR codes has it's time synced correctly with a time server? - if it's out by a few seconds it might make the expiry time of the qr codes shorter than it should be.

**Comment by danmarsden** [comment]
the other thing you could consider is increasing the "rotateqrcodeexpirymargin" setting at the admin level to make the qrcodes valid for a little bit longer too.

I'd be interested to know what qrcodeinterval you are using too.

**Comment by mdorferf** [comment]
Dear Dan,
Thank you for the changes and sorry for my delayed answer!

The changes are very welcome from our side. We think it's good to hide the password hint. It doesn't make sense in this place. Thanks also for the help page adjustments! It would be good to have an area for further tips, such as clearing the cache, or simply trying again with another browser. This could maybe eliminate some sources of error. 

We analyzed the issue with our developers, but we still have the problem. Attached you will find our admin parameters. 
As we understand it, we have 30+45 seconds to capture attendance in the best case, and 1+45 in the shortest case, correct? Do you have further recommendations for our admin parameters?

![Image](https://github.com/user-attachments/assets/8188e06e-f086-47aa-97b5-85648d83bef7)

We also did tests together with significantly changed local time of the client (laptop) – we were only about 700 milliseconds off. But still having problems with the session, see error message

![Image](https://github.com/user-attachments/assets/bbb9e53d-125f-4ee5-8778-b5caf3e1d408)

Further thoughts from our devs:
•	Development according to RFC 6238, suggestion: only save the seed in the database and calculate the current value on the client. Only one request necessary and we can show correct QR codes indefinitely without further AJAX calls, we only need the seed and the current time. Create QR codes on the client via a JS lib. Verification on the server only needs the seed value and current time.
•	For server side implementation look at Moodle's libraries for multi-factor authentication for TOTP (see admin/tool/mfa/factor/totp/extlib/OTPHP) 

As we are aware of the development effort, we can consider sharing the costs. 
Thank you in advance for your help!

**Comment by danmarsden** [comment]
yeah - I'd be keen to change out how a lot of that QR code rotation works - it actually dates back to a GSOC project and it's not quite the way I wanted to have it finished off, but provided a good improvement which has served us pretty well most of the time.

I don't currently have a way to crowd-fund improvements to any of the plugins I work on and most crowd-funding projects I've seen related to Moodle haven't been all that succesful - your suggestion does seem like a good idea, however it's not a small piece of work and might be hard to get other people interested in funding the development.

Always happy to receive pull requests, and feel free to reach out privately if you would like to independently fund specific improvements.

thanks!
